src/backend/distributed/connection/locally_reserved_shared_connections.c

```diff
@@ -342,62 +360,106 @@ EnsureConnectionPossibilityForNodeList(List *nodeList)
-		 *
-		 * Note that shared connection stats operate on the shared memory, and we
-		 * pre-allocate all the necessary memory. In other words, it would never
-		 * throw out of memory error.
+		 * The same user has already an active connection for the node. It
+		 * means that the execution can use the same connection, so reservation
+		 * is not necessary.
 		 */
-		bool found = false;
-		ReservedConnectionHashEntry *hashEntry =
-			AllocateOrGetReservedConectionEntry(workerNode->workerName,
-												workerNode->workerPort,
-												userId, MyDatabaseId, &found);
+		return true;
+	}
 
-		if (found)
-		{
-			/*
-			 * We have already reserved a connection for this user and database
-			 * on the worker. We only allow a single reservation per
-			 * transaction block. The reason is that the earlier command (either in
-			 * a transaction block or a function call triggered by a single command)
-			 * was able to reserve or establish a connection. That connection is
-			 * guranteed to be avaliable for us.
-			 */
-			continue;
-		}
+	/*
+	 * We are trying to be defensive here by ensuring that the required hash
+	 * table entry can be allocated. The main goal is that we don't want to be
+	 * in a situation where shared connection counter is incremented but not
+	 * the local reserved counter due to out-of-memory.
+	 *
+	 * Note that shared connection stats operate on the shared memory, and we
+	 * pre-allocate all the necessary memory. In other words, it would never
+	 * throw out of memory error.
+	 */
+	bool found = false;
+	ReservedConnectionHashEntry *hashEntry =
+		AllocateOrGetReservedConectionEntry(workerNode->workerName,
+											workerNode->workerPort,
+											userId, MyDatabaseId, &found);
+
+	if (found)
+	{
+		/*
+		 * We have already reserved a connection for this user and database
+		 * on the worker. We only allow a single reservation per
+		 * transaction block. The reason is that the earlier command (either in
+		 * a transaction block or a function call triggered by a single command)
+		 * was able to reserve or establish a connection. That connection is
+		 * guranteed to be available for us.
+		 */
+		return true;
+	}
 
+	if (waitForConnection)
+	{
 		/*
 		 * Increment the shared counter, we may need to wait if there are
 		 * no space left.
 		 */
 		WaitLoopForSharedConnection(workerNode->workerName, workerNode->workerPort);
+	}
+	else
+	{
+		bool incremented =
+			TryToIncrementSharedConnectionCounter(workerNode->workerName,
+												  workerNode->workerPort);
 
-		/* locally mark that we have one connection reserved */
-		hashEntry->usedReservation = false;
+		if (!incremented)
+		{
+			/*
+			 * We could not reserve a connection. First, remove the entry from the
+			 * hash. The reason is that we allow single reservation per transaction
+			 * block and leaving the entry in the hash would be qualified as there is a
+			 * reserved connection to the node.
+			 */
+			bool foundForRemove = false;
```

Comment:
This is a little unexpected part of the changes. We had to `HASH_REMOVE ` because the meaning of having a hash entry that Citus reserved a connection to the node. Looking for ways to skip such a behavior but doesn't seem trivial avoid due to:

```C
	if (found)
	{
		/*
		 * We have already reserved a connection for this user and database
		 * on the worker. We only allow a single reservation per
		 * transaction block. The reason is that the earlier command (either in
		 * a transaction block or a function call triggered by a single command)
		 * was able to reserve or establish a connection. That connection is
		 * guranteed to be available for us.
		 */
		return true;
	}
```